alibi_detect/od/base.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from typing import Dict, Any
+
+from typing_extensions import Protocol, runtime_checkable
+import numpy as np
+
+from alibi_detect.base import BaseDetector
+
+
+class OutlierDetector(BaseDetector, ABC):
+    threshold_inferred = False
+    ensemble = False
+
+    def __init__(self):
+        """ Base class for outlier detection algorithms."""
+        super().__init__()
+        self.meta['online'] = False
+        self.meta['detector_type'] = 'outlier'
+
+    @abstractmethod
+    def fit(self, X: np.ndarray) -> None:
+        """
+        Fit outlier detector to data.
+
+        Parameters
+        ----------
+        X
+            Reference data.
+        """
+        pass
+
+    @abstractmethod
+    def score(self, X: np.ndarray) -> np.ndarray:
+        """
+        Compute anomaly scores of the instances in X.
+
+        Parameters
+        ----------
+        X
+            Data to score.
+
+        Returns
+        -------
+        Anomaly scores. The higher the score, the more anomalous the instance.
+        """
+        pass
+
+    @abstractmethod
+    def infer_threshold(self, X: np.ndarray, fpr: float) -> None:
+        """
+        Infer the threshold for the outlier detector.
+
+        Parameters
+        ----------
+        X
+            Reference data.
+        fpr
+            False positive rate used to infer the threshold.
+        """
+        pass
+
+    @abstractmethod
+    def predict(self, X: np.ndarray) -> Dict[str, Any]:
```

Comment:
I did use the `dataclass` approach here because it would be a significant break from the rest of alibi-detect and I thought it would be strange for users if drift detectors return dictionaries whereas outlier detectors return dataclasses. Shall I open an issue for replacing throughout though?

Reply:
Yeh fair point. Nice idea to open an issue about it. There could already be a related issue, I can't remember!

Reply:
Created issue [here](https://github.com/SeldonIO/alibi-detect/issues/762)